### PR TITLE
docs: add readthedocs build using doxygen

### PR DESCRIPTION
**Pull request by ReenigneArcher**
## Description

Adds doxygen docs, with a docs target, and readthedocs configuration.

Todo:
- [x] all builds pass
  - [x] Linux
  - [x] macOS
  - [x] Windows
- [x] add project icon/logo
- [x] fix logo in header.html
- [x] Add regular markdown files to doxygen build
  - [x] convert README.rst to README.md

### Screenshot

Preview: https://lizardbyte--71.org.readthedocs.build/projects/libdisplaydevice/71/

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Dependency update (updates to dependencies)
- [x] Documentation update (changes to documentation)
- [x] Repository update (changes to repository files, e.g. `.github/...`)

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated the in code docstring/documentation-blocks for new or existing methods/components

### CODE REVIEW

File: src/common/json_serializer.cpp [hunk 1]
```diff
@@ -1,3 +1,7 @@
+/**
+ * @file src/common/json_serializer.cpp
+ * @brief Definitions for JSON serialization.
```
Comment (FrogTheFrog) [inline src/common/json_serializer.cpp:3]:
```suggestion
 * @brief Definitions for private JSON serialization helpers.
```

File: src/common/types.cpp [hunk 2]
```diff
@@ -1,3 +1,7 @@
+/**
+ * @file src/common/types.cpp
+ * @brief Definitions for display device types.
```
Comment (FrogTheFrog) [inline src/common/types.cpp:3]:
```suggestion
 * @brief Definitions for common display device types.
```

File: src/windows/include/display_device/windows/detail/json_serializer.h [hunk 3]
```diff
@@ -1,3 +1,7 @@
+/**
+ * @file src/windows/include/display_device/windows/detail/json_serializer.h
+ * @brief Declarations for JSON serialization of display device structs.
```
Comment (FrogTheFrog) [inline src/windows/include/display_device/windows/detail/json_serializer.h:3]:
```suggestion
 * @brief Declarations for private JSON serialization helpers (Windows-only).
```

File: src/windows/include/display_device/windows/json.h [hunk 4]
```diff
@@ -1,3 +1,7 @@
+/**
+ * @file src/windows/include/display_device/windows/json.h
+ * @brief Declarations for JSON converters for Windows-specific types.
```
Comment (FrogTheFrog) [inline src/windows/include/display_device/windows/json.h:3]:
```suggestion
 * @brief Declarations for JSON conversion functions (Windows-only).
```

File: src/windows/include/display_device/windows/settings_utils.h [hunk 5]
```diff
@@ -1,3 +1,7 @@
+/**
+ * @file src/windows/include/display_device/windows/settings_utils.h
+ * @brief Declarations for shared "utility-level" code for Settings.
```
Comment (FrogTheFrog) [inline src/windows/include/display_device/windows/settings_utils.h:3]:
```suggestion
 * @brief Declarations for settings related utility functions.
```

File: src/windows/include/display_device/windows/win_api_utils.h [hunk 6]
```diff
@@ -1,3 +1,7 @@
+/**
+ * @file src/windows/include/display_device/windows/win_api_utils.h
+ * @brief Declarations for Windows API utility functions.
```
Comment (FrogTheFrog) [inline src/windows/include/display_device/windows/win_api_utils.h:3]:
```suggestion
 * @brief Declarations for lower level Windows API utility functions.
```

File: src/windows/json_serializer.cpp [hunk 7]
```diff
@@ -1,3 +1,7 @@
+/**
+ * @file src/windows/json_serializer.cpp
+ * @brief Definitions for JSON serialization of display device structs.
```
Comment (FrogTheFrog) [inline src/windows/json_serializer.cpp:3]:
```suggestion
 * @brief Definitions for private JSON serialization helpers (Windows-only).
```

File: src/windows/json.cpp [hunk 8]
```diff
@@ -1,3 +1,7 @@
+/**
+ * @file src/windows/json.cpp
+ * @brief Definitions for JSON converters for Windows-specific types.
```
Comment (FrogTheFrog) [inline src/windows/json.cpp:3]:
```suggestion
 * @brief Definitions for JSON conversion functions (Windows-only).
```

File: src/windows/settings_manager_apply.cpp [hunk 9]
```diff
@@ -1,3 +1,7 @@
+/**
+ * @file src/windows/settings_manager_apply.cpp
+ * @brief Definitions for the SettingsManager (apply).
```
Comment (FrogTheFrog) [inline src/windows/settings_manager_apply.cpp:3]:
```suggestion
 * @brief Definitions for the methods for applying settings in SettingsManager.
```

File: src/windows/settings_manager_general.cpp [hunk 10]
```diff
@@ -1,3 +1,7 @@
+/**
+ * @file src/windows/settings_manager_general.cpp
+ * @brief Definitions for the SettingsManager (general).
```
Comment (FrogTheFrog) [inline src/windows/settings_manager_general.cpp:3]:
```suggestion
 * @brief Definitions for the leftover (general) methods in SettingsManager.
```

File: src/windows/settings_manager_revert.cpp [hunk 11]
```diff
@@ -1,3 +1,7 @@
+/**
+ * @file src/windows/settings_manager_revert.cpp
+ * @brief Definitions for the SettingsManager (revert).
```
Comment (FrogTheFrog) [inline src/windows/settings_manager_revert.cpp:3]:
```suggestion
 * @brief Definitions for the methods for reverting settings in SettingsManager..
```

File: src/windows/settings_utils.cpp [hunk 12]
```diff
@@ -1,3 +1,7 @@
+/**
+ * @file src/windows/settings_utils.cpp
+ * @brief Definitions for shared "utility-level" code for Settings.
```
Comment (FrogTheFrog) [inline src/windows/settings_utils.cpp:3]:
```suggestion
 * @brief Definitions for settings related utility functions.
```

File: src/windows/win_api_utils.cpp [hunk 13]
```diff
@@ -1,3 +1,7 @@
+/**
+ * @file src/windows/win_api_utils.cpp
+ * @brief Definitions for Windows API utility functions.
```
Comment (FrogTheFrog) [inline src/windows/win_api_utils.cpp:3]:
```suggestion
 * @brief Definitions for lower level Windows API utility functions.
```

File: src/windows/win_display_device_general.cpp [hunk 14]
```diff
@@ -1,3 +1,7 @@
+/**
+ * @file src/windows/win_display_device_general.cpp
+ * @brief Definitions for the WinDisplayDevice (general).
```
Comment (FrogTheFrog) [inline src/windows/win_display_device_general.cpp:3]:
```suggestion
 * @brief Definitions for the leftover (general) methods in WinDisplayDevice.
```

File: src/windows/win_display_device_hdr.cpp [hunk 15]
```diff
@@ -1,3 +1,7 @@
+/**
+ * @file src/windows/win_display_device_hdr.cpp
+ * @brief Definitions for the WinDisplayDevice (hdr).
```
Comment (FrogTheFrog) [inline src/windows/win_display_device_hdr.cpp:3]:
```suggestion
 * @brief Definitions for the HDR related methods in WinDisplayDevice.
```

File: src/windows/win_display_device_modes.cpp [hunk 16]
```diff
@@ -1,3 +1,7 @@
+/**
+ * @file src/windows/win_display_device_modes.cpp
+ * @brief Definitions for the WinDisplayDevice (modes).
```
Comment (FrogTheFrog) [inline src/windows/win_display_device_modes.cpp:3]:
```suggestion
 * @brief Definitions for the display mode related methods in WinDisplayDevice.
```

File: src/windows/win_display_device_primary.cpp [hunk 17]
```diff
@@ -1,3 +1,7 @@
+/**
+ * @file src/windows/win_display_device_primary.cpp
+ * @brief Definitions for the WinDisplayDevice (primary).
```
Comment (FrogTheFrog) [inline src/windows/win_display_device_primary.cpp:3]:
```suggestion
 * @brief Definitions for the "primary device" related methods in WinDisplayDevice.
```

File: src/windows/win_display_device_topology.cpp [hunk 18]
```diff
@@ -1,3 +1,7 @@
+/**
+ * @file src/windows/win_display_device_topology.cpp
+ * @brief Definitions for the WinDisplayDevice (topology).
```
Comment (FrogTheFrog) [inline src/windows/win_display_device_topology.cpp:3]:
```suggestion
 * @brief Definitions for the topology related methods in WinDisplayDevice.
```

File: docs/CMakeLists.txt [hunk 19]
```diff
@@ -0,0 +1,28 @@
+cmake_minimum_required(VERSION 3.19)
+
```
Comment (ReenigneArcher) [inline docs/CMakeLists.txt:2]:
```suggestion
```

File: docs/CMakeLists.txt [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+if(DEFINED ENV{READTHEDOCS})
+    set(LDD_DOXYGEN_BUILD_DIR $ENV{READTHEDOCS_OUTPUT})
+    set(LDD_PROJECT_VERSION $ENV{READTHEDOCS_VERSION})
+else()
+    set(LDD_DOXYGEN_BUILD_DIR "${CMAKE_CURRENT_BINARY_DIR}/build")
+    set(LDD_PROJECT_VERSION ${PROJECT_VERSION})
+endif()
+
+# doxygen fails to create the output directory if it does not exist, so let's handle it here
+file(MAKE_DIRECTORY ${LDD_DOXYGEN_BUILD_DIR})
```
Comment (ReenigneArcher) [inline docs/CMakeLists.txt:17]:
```suggestion
file(MAKE_DIRECTORY "${LDD_DOXYGEN_BUILD_DIR}/html")
```

File: docs/CMakeLists.txt [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+# build docs
+add_custom_target(docs ALL
+        WORKING_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}"
+        COMMENT "Building Doxygen documentation"
+        COMMAND ${CMAKE_COMMAND} -E env
+            LDD_DOXYGEN_BUILD_DIR=${LDD_DOXYGEN_BUILD_DIR}
+            LDD_PROJECT_VERSION=${LDD_PROJECT_VERSION}
+            ${DOXYGEN_EXECUTABLE} Doxyfile
+        VERBATIM
+)
```
Comment (ReenigneArcher) [inline docs/CMakeLists.txt:26]:
```suggestion
# build docs
add_custom_target(docs ALL
        WORKING_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}"
        COMMENT "Building Doxygen documentation"
        # doxygen fails to create the output directory if it does not exist, so let's handle it here
        COMMAND ${CMAKE_COMMAND} -E make_directory "${LDD_DOXYGEN_BUILD_DIR}/html"
        COMMAND ${CMAKE_COMMAND} -E env
            LDD_DOXYGEN_BUILD_DIR=${LDD_DOXYGEN_BUILD_DIR}
            LDD_PROJECT_VERSION=${LDD_PROJECT_VERSION}
            ${DOXYGEN_EXECUTABLE} Doxyfile
        VERBATIM
)
```

File: docs/CMakeLists.txt [hunk 22]
```diff
@@ -0,0 +1,25 @@
...
+        WORKING_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}"
+        COMMENT "Building Doxygen documentation"
+        # doxygen fails to create the output directory if it does not exist, so let's handle it here
+        COMMAND ${CMAKE_COMMAND} -E make_directory "${LDD_DOXYGEN_BUILD_DIR}/html"
+        COMMAND ${CMAKE_COMMAND} -E env
+            LDD_DOXYGEN_BUILD_DIR=${LDD_DOXYGEN_BUILD_DIR}
+            LDD_PROJECT_VERSION=${LDD_PROJECT_VERSION}
+            ${DOXYGEN_EXECUTABLE} Doxyfile
+        VERBATIM
+)
```
Comment (ReenigneArcher) [inline docs/CMakeLists.txt:25]:
```suggestion
# create build directories, as doxygen fails to create it in some cases?
file(MAKE_DIRECTORY "${LDD_DOXYGEN_BUILD_DIR}/html")

# build docs
add_custom_target(docs ALL
        WORKING_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}"
        COMMENT "Building Doxygen documentation"
        COMMAND ${CMAKE_COMMAND} -E env
            LDD_DOXYGEN_BUILD_DIR=${LDD_DOXYGEN_BUILD_DIR}
            LDD_PROJECT_VERSION=${LDD_PROJECT_VERSION}
            ${DOXYGEN_EXECUTABLE} Doxyfile
        VERBATIM
)
```

**Review by ReenigneArcher — COMMENTED** [review]
See here: https://github.com/LizardByte/Sunshine/blob/459c604803742d0047c4a15ffe5e5649f1c9bdc4/.github/workflows/CI.yml#L1078-L1101

File: docs/CMakeLists.txt [hunk 23]
```diff
@@ -0,0 +1,29 @@
...
+# build docs
+add_custom_target(docs ALL
+        WORKING_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}"
+        COMMENT "Building Doxygen documentation"
+        COMMAND ${CMAKE_COMMAND} -E env
+            LDD_DOXYGEN_BUILD_DIR=${LDD_DOXYGEN_BUILD_DIR}
+            LDD_PROJECT_VERSION=${LDD_PROJECT_VERSION}
+            ${DOXYGEN_EXECUTABLE} Doxyfile
+        VERBATIM
+)
```
Comment (ReenigneArcher) [inline docs/CMakeLists.txt:29]:
```suggestion
# define variables based on whether we are building on readthedocs
if(DEFINED ENV{READTHEDOCS})
    set(DOXYGEN_BUILD_DIR_CMAKE $ENV{READTHEDOCS_OUTPUT})
    set(DOXYGEN_PROJECT_VERSION $ENV{READTHEDOCS_VERSION})
else()
    set(DOXYGEN_BUILD_DIR_CMAKE "${CMAKE_CURRENT_BINARY_DIR}/build")
    set(DOXYGEN_PROJECT_VERSION ${PROJECT_VERSION})
endif()

# create build directories, as doxygen fails to create it in some cases?
file(MAKE_DIRECTORY "${DOXYGEN_BUILD_DIR_CMAKE}/html")

# convert to native path, so doxygen doesn't get confused on Windows
file(TO_NATIVE_PATH "${DOXYGEN_BUILD_DIR_CMAKE}" DOXYGEN_BUILD_DIR_NATIVE)

# build docs
add_custom_target(docs ALL
        WORKING_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}"
        COMMENT "Building Doxygen documentation"
        COMMAND ${CMAKE_COMMAND} -E env
            LDD_DOXYGEN_BUILD_DIR=${DOXYGEN_BUILD_DIR_NATIVE}
            LDD_PROJECT_VERSION=${DOXYGEN_PROJECT_VERSION}
            ${DOXYGEN_EXECUTABLE} Doxyfile
        VERBATIM
)
```

File: docs/CMakeLists.txt [hunk 24]
```diff
@@ -0,0 +1,29 @@
...
+# build docs
+add_custom_target(docs ALL
+        WORKING_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}"
+        COMMENT "Building Doxygen documentation"
+        COMMAND ${CMAKE_COMMAND} -E env
+            LDD_DOXYGEN_BUILD_DIR=${DOXYGEN_BUILD_DIR_NATIVE}
+            LDD_PROJECT_VERSION=${DOXYGEN_PROJECT_VERSION}
+            ${DOXYGEN_EXECUTABLE} Doxyfile
+        VERBATIM
+)
```
Comment (ReenigneArcher) [inline docs/CMakeLists.txt:29]:
```suggestion
# convert to relative path, so doxygen doesn't get confused on Windows
file(RELATIVE_PATH DOXYGEN_BUILD_DIR_RELATIVE "${CMAKE_CURRENT_SOURCE_DIR}" "${DOXYGEN_BUILD_DIR_CMAKE}")

# build docs
add_custom_target(docs ALL
        WORKING_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}"
        COMMENT "Building Doxygen documentation"
        COMMAND ${CMAKE_COMMAND} -E env
            LDD_DOXYGEN_BUILD_DIR=${DOXYGEN_BUILD_DIR_RELATIVE}
            LDD_PROJECT_VERSION=${DOXYGEN_PROJECT_VERSION}
            ${DOXYGEN_EXECUTABLE} Doxyfile
        VERBATIM
)
```

**Pull request merged**